Add http support, and add url parameter to the client constructor

tattish opened the pull request:
Hi,

I would like to contribute to this project, i already did most of the changes that you refactored in the last month, like json output or subcommands or custom url parsing, but i decided to drop it away, and follow your branch. My plan is to introduce mostly the full features for the API. (I use the on-prem setup of MMS with version 1.5.3 that is why the first step was to make the url passing working and as our test setup is http to make the ssl optional. Next will try to introduce multiversion support for 1.6 and 1.5 MMS API compatible calls. ) I hope we can work together further, any input is very welcomed, and many thanks for creating this project.

This pull request is about to add the http support , i changed a little the client to wrap the protocol out of the url, and if it is not https, turn off the ssl flag. As well added the url parameter as parameter to the client constructor call in the cli implementation.

Regards, 
@t

kris-lab commented:
Hi,

Thanks a lot for contributing to this project. The goal of initially simple client was to coverage our needs in area of mongo cluster automations. The main goals have been achieved so I am really happy if more developers would like to continue on that.

We will fully support and maintenance all work requirements on that project in reasonable timeframe.

In next few days I will also talk to involved developers about the future of this repo; will be back to you with news asap.

Thanks again for current PR. I will have a look, test and merge soon as possible.

Regards
@kris-lab

tomaszdurka commented on lib/mms/cli.rb:
```diff
@@ -68,12 +68,11 @@ def parse_user_home_config
       # @return [MMS::Agent]
       def agent
-        @client = MMS::Client.new(@config.username, @config.apikey)
+        @client = MMS::Client.new(@config.username, @config.apikey, url=@config.apiurl)
```
Didn't  you set it somewhere? `agent.set_apiurl`

kris-lab replied:
Could you please remove `url=`? I believe it is not needed?

tattish replied:
@kris-lab : I can remove the url= it is not needed , it is due to a bad habit on my side, i like better to see as what is the destination param, but i understand that breaks the semantics what was used so far, 
@tomaszdurka : As the url does not change during the run, i thought it is better to provide it here in the constructor, specially it was already in as a parameter.

tomaszdurka commented on lib/mms/client.rb:
```diff
@@ -16,7 +16,7 @@ def initialize(username = nil, apikey = nil, url = nil)
 
     # @param [String] path
-    # @return [Hash]
+    # @return [Hash]     
```
Whitespace.

tattish replied:
That certainly just a mistake. :)

tomaszdurka commented on lib/mms/client.rb:
```diff
@@ -29,7 +29,11 @@ def post(path, data)
-
+    
+    def _get_ssl()
+      return (@url.split(":")[0]=="https")
```
Maybe rather `URI.parse(@url).scheme === 'https'`

tomaszdurka replied:
Oh wait, let's create one method `_request(method, path, username, password)`?

tattish replied:
I may will introduce DELETE and PATCH calls as well, as the API follows this scheme, so the _request seems a good idea

tomaszdurka commented:
Hey @tattish sorry about bit detailed review which covers more than you try to change within your pull-request. @kris-lab asked me for short look and I actually over-reviewed your pull-request (as I thought it's coming from @kris-lab).

I think your change makes sense. I would maybe inline that `_get_ssl` into `_get`/`_post` methods as you have `URI::Http/s` instance there already and can use `.scheme` to detect `https`. Another reason to inline is as mentioned we aim to refactor `_post`/`_get` methods some time soon to have common base (remove duplication).

Thanks.

tattish commented:
I would more like to say thanks for the detailed review. As i wrote above, it will be needed to introduce DELETE, PATCH maybe even PUT for some API calls, i am totally ok with inline the http/s check.

tomaszdurka commented:
Hey @tattish good job! 
Only a few notes :)

Let's change signature of `_request` method so `method` param is required and moved to the front:
```ruby
def _request(method, path, username, password, data = nil)
```

Let's also simplify `method` param. I know you've made it Hash so you can pass name and http request class, but class might be enough.

In lib/mms/client.rb#_request() we could
```ruby
raise 'Invalid method' unless method.kind_of? Class and method < Net::HTTPRequest
req = method.new(uri.request_uri, {'Content-Type' => 'application/json'})
method_name = method.class.name.split('::').last
auth = digest_auth.auth_header(uri, res['WWW-Authenticate'], method_name)
```

tattish commented:
Hey @tomaszdurka ,
I added the proposed changes, and to avoid confusion i changed the param name from method to http_method. also added an upcase to the method name derivation.

tomaszdurka commented:
Looks good to me, @kris-lab please merge if you feel right about it.

kris-lab commented:
lgtm. will be released in `0.1.2`
